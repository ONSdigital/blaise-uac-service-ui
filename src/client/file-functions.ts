import axios, { AxiosRequestConfig } from "axios";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { Datas } from "react-csv-downloader/dist/esm/lib/csv";

function axiosConfig(): AxiosRequestConfig {
    const authManager = new AuthManager();
    return {
        headers: authManager.authHeader()
    };
}

export async function importUacsFromFile(file: File | undefined): Promise<number> {
    if (file === undefined) {
        throw new Error("file was not supplied");
    }

    const allowedFileTypes = /(\.csv)$/i;
    if (!allowedFileTypes.exec(file.name)) {
        throw new Error("File format is not CSV");
    }

    const data = new FormData();
    data.append("file", file);

    return axios.post("/api/v1/uac/import", data, axiosConfig())
        .then((response) => {
            console.log("import-file-function - true");
            console.log(response.data);
            return response.data?.uacs_imported;
        })
        .catch((error) => {
            console.log("import-file-function - false");
            console.error(`Something went wrong importing uacs ${error}`);
            throw error;
        });
}

export async function generateUacCodesForSampleFile(instrumentName: string | undefined, file: File | undefined): Promise<boolean> {
    if (instrumentName === undefined) {
        throw new Error("Questionnaire name was not supplied");
    }

    if (file === undefined) {
        throw new Error("file was not supplied");
    }

    const data = new FormData();
    data.append("fileName", getFileName(instrumentName));
    data.append("file", file);

    return axios.post(`/api/v1/instrument/${instrumentName}/uac/sample`, data, axiosConfig())
        .then(() => {
            console.log("file-functions - true");
            return true;
        })
        .catch((error) => {
            console.log("file-functions - false");
            console.error(`Something went wrong in calling generate UAC endpoint ${error}`);
            throw error;
        });
}

export async function getSampleFileWithUacCodes(instrumentName: string | undefined, fileName: string | undefined): Promise<Datas> {
    if (instrumentName === undefined) {
        throw new Error("Questionnaire name was not supplied");
    }

    if (fileName === undefined) {
        throw new Error("file name was not supplied");
    }

    const response = await axios.get(`/api/v1/instrument/${instrumentName}/uac/sample/${fileName}`, axiosConfig());

    return response.data;
}

export async function sampleFileAlreadyExists(instrumentName: string | undefined): Promise<boolean> {
    if (instrumentName === undefined) {
        throw new Error("Questionnaire name was not supplied");
    }

    const fileName = getFileName(instrumentName);
    return axios.get(`/api/v1/file/${fileName}/exists`, axiosConfig())
        .then((response) => response.data === true)
        .catch((error) => {
            console.error(error);
            return false;
        });
}

export async function getListOfInstrumentsWhichHaveExistingSampleFiles(): Promise<string[]> {
    const response = await axios.get("/api/v1/instruments", axiosConfig());

    return response.data;
}

export function getFileName(instrumentName: string): string {
    return `${instrumentName}.csv`;
}
