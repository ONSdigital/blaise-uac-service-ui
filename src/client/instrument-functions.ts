import axios from "axios";

export async function getInstrumentsWithExistingUacCodes(): Promise<string[]> {
    const response = await axios.get("/api/v1/instruments2");
    console.log("List of instruments: ", response);

    return [];
}
