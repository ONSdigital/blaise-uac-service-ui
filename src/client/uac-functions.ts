import axios from "axios";

export async function getInstrumentsWithExistingUacCodes() {
    const response = await axios.get("/api/v1/uac");
    console.log("Uac meh response", response);

    return response;
}
