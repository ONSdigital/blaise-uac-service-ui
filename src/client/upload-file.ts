import axios from "axios";

export async function uploadFile(instrumentName: string | undefined, file: File | undefined) {
    if (instrumentName === undefined || instrumentName === null) {
        console.error("Instrument name not supplied");
        return null;
    }

    if (file === undefined || file === null) {
        console.error("File not supplied");
        return null;
    }

    const data = new FormData();
    data.append("instrumentName", instrumentName);
    data.append("file", file);

    const config = {headers: {"Content-Type": "multipart/form-data"}};

    await axios.post("/api/v1/upload", data, config);
}
