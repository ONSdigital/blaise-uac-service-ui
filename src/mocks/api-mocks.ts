import { Questionnaire } from "blaise-api-node-client";
import { InstrumentUacDetails } from "blaise-uac-service-node-client";
import { QuestionnaireWithDisabledUacs } from "../models/model";

export const instrumentNames: string[] = [
    "OPN2101A",
    "DST1345A",
    "DST5643A",
];

export const questionnaireList: Questionnaire[] = [{
    name: "OPN2101A",
    serverParkName: "gusty",
    installDate: "2021-01-15T14:41:29.4399898+00:00",
    status: "Active",
    dataRecordCount: 0,
    hasData: false,
    active: false,
    fieldPeriod: "January 2021",
    blaiseVersion: "5.9.9.2735"
}, {
    name: "OPN2007T",
    serverParkName: "gusty",
    installDate: "2021-01-15T15:18:40.1503617+00:00",
    status: "Active",
    dataRecordCount: 10,
    hasData: true,
    active: true,
    fieldPeriod: "July 2020",
    blaiseVersion: "5.9.9.2735"

}, {
    name: "OPN2004A",
    serverParkName: "gusty",
    installDate: "2021-01-15T15:26:43.4233454+00:00",
    status: "Failed",
    dataRecordCount: 0,
    hasData: false,
    active: false,
    fieldPeriod: "April 2020",
    blaiseVersion: "5.9.9.2735"

}];

export const questionnaireListForEnableUacsMock: Questionnaire[] = [{
    name: "LMS2209_EM1",
    serverParkName: "gusty",
    installDate: "2021-01-15T14:41:29.4399898+00:00",
    status: "Active",
    dataRecordCount: 0,
    hasData: false,
    active: false,
    fieldPeriod: "January 2021",
    blaiseVersion: "5.9.9.2735"
}, {
    name: "LMS2207_HO1",
    serverParkName: "gusty",
    installDate: "2021-01-15T15:18:40.1503617+00:00",
    status: "Active",
    dataRecordCount: 10,
    hasData: true,
    active: true,
    fieldPeriod: "July 2020",
    blaiseVersion: "5.9.9.2735"

}];

export const disabledUacCodesForQuestionnaireMock: InstrumentUacDetails = {
    "803920": {
        "instrument_name": "lms2209_em1",
        "case_id": "803920",
        "uac_chunks": { "uac1": "1002", "uac2": "2293", "uac3": "8976" },
        "full_uac": "100222938976",
        "disabled": "true"
    },
    "804138": {
        "instrument_name": "lms2209_em1",
        "case_id": "804138",
        "uac_chunks": { "uac1": "1002", "uac2": "6087", "uac3": "6564" },
        "full_uac": "100260876564",
        "disabled": "true"
    },
    "907195": {
        "instrument_name": "lms2209_em1",
        "case_id": "907195",
        "uac_chunks": { "uac1": "1004", "uac2": "6119", "uac3": "7282" },
        "full_uac": "100461197282",
        "disabled": "true"
    }
};

export const questionnaireWithDisabledUacsMock: QuestionnaireWithDisabledUacs =
{
    "questionnaireName": "LMS2209_EM1",
    "disabledUacs": [
        {
            "case_id": "803920",
            "uac": "100222938976"
        },
        {
            "case_id": "804138",
            "uac": "100260876564"
        },
        {
            "case_id": "907195",
            "uac": "100461197282"
        },

    ]
};

export const questionnaireWithOneDisabledUacMock: QuestionnaireWithDisabledUacs =
{
    "questionnaireName": "LMS2209_EM1",
    "disabledUacs": [
        {
            "case_id": "803920",
            "uac": "100222938976"
        }

    ]
};