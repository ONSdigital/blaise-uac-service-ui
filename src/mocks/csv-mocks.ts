import { InstrumentUacDetailsByCaseId } from "blaise-uac-service-node-client";
import { Datas } from "react-csv-downloader/dist/esm/lib/csv";

export const validUACImportCsv = `Full_UAC
123412341234
432143214321
678967896789
987698769876
`;

export const validSampleCsv = `serial_number,Name,Phone Number,Email
100000001,Homer Simpson,5551234422,homer@springfield.com
100000002,Seymour Skinner,1235663322,a@b.c
100000003,Bart Simpson,2675465026,bart@spring.field
`;

export const validSampleCsvWithExistingUacColumns = `serial_number,Name,UAC1,UAC2,UAC3,Phone Number,Email
100000001,Homer Simpson,,,,5551234422,homer@springfield.com
100000002,Seymour Skinner,,,,1235663322,a@b.c
100000003,Bart Simpson,,,,2675465026,bart@spring.field
`;

export const validSampleCsvWithExistingLowercaseUacColumns = `serial_number,Name,uac1,uac2,uac3,Phone Number,Email
100000001,Homer Simpson,,,,5551234422,homer@springfield.com
100000002,Seymour Skinner,,,,1235663322,a@b.c
100000003,Bart Simpson,,,,2675465026,bart@spring.field
`;

export const validSampleCsvWithExistingMixedCaseUacColumns = `serial_number,Name,UaC1,UaC2,UaC3,Phone Number,Email
100000001,Homer Simpson,,,,5551234422,homer@springfield.com
100000002,Seymour Skinner,,,,1235663322,a@b.c
100000003,Bart Simpson,,,,2675465026,bart@spring.field
`;

export const validSampleCsvWithExistingUacEntries = `serial_number,Name,UAC1,UAC2,UAC3,Phone Number,Email
100000001,Homer Simpson,3454,5453,5353,5551234422,homer@springfield.com
100000002,Seymour Skinner,8786,2213,3343,1235663322,a@b.c
100000003,Bart Simpson,2242,5543,7676,2675465026,bart@spring.field
`;

export const invalidSampleCsv = `serial_number,Phone Number,Email
100000001,Homer Simpson,5551234422,homer@springfield.com
100000002,Seymour Skinner,1235663322,a@b.c
100000003,Bart Simpson,2675465026,bart@spring.field
`;


export const matchedInstrumentUacDetails: InstrumentUacDetailsByCaseId = {
    "100000001": {
        instrument_name: "dst2106a",
        case_id: "100000001",
        uac_chunks: {
            uac1: "0009",
            uac2: "7565",
            uac3: "3827"
        },
        full_uac: "000975653827"
    },
    "100000002": {
        instrument_name: "dst2106a",
        case_id: "100000002",
        uac_chunks: {
            uac1: "3453",
            uac2: "6545",
            uac3: "4564"
        },
        full_uac: "345365454564"
    },
    "100000003": {
        instrument_name: "dst2106a",
        case_id: "100000003",
        uac_chunks: {
            uac1: "9789",
            uac2: "7578",
            uac3: "5367"
        },
        full_uac: "978975785367"
    }
};

export const matchedInstrumentUac16Details: InstrumentUacDetailsByCaseId = {
    "100000001": {
        instrument_name: "dst2106a",
        case_id: "100000001",
        uac_chunks: {
            uac1: "0009",
            uac2: "7565",
            uac3: "3827",
            uac4: "7512"
        },
        full_uac: "0009756538277512"
    },
    "100000002": {
        instrument_name: "dst2106a",
        case_id: "100000002",
        uac_chunks: {
            uac1: "3453",
            uac2: "6545",
            uac3: "4564",
            uac4: "3213"
        },
        full_uac: "3453654545643213"
    },
    "100000003": {
        instrument_name: "dst2106a",
        case_id: "100000003",
        uac_chunks: {
            uac1: "9789",
            uac2: "7578",
            uac3: "5367",
            uac4: "8765"
        },
        full_uac: "9789757853678765"
    }
};

export const unMatchedInstrumentUacDetails: InstrumentUacDetailsByCaseId = {
    "100000997": {
        instrument_name: "dst2106a",
        case_id: "100000997",
        uac_chunks: {
            uac1: "0009",
            uac2: "7565",
            uac3: "3827"
        },
        full_uac: "000975653827"
    },
    "100000998": {
        instrument_name: "dst2106a",
        case_id: "100000998",
        uac_chunks: {
            uac1: "3453",
            uac2: "6545",
            uac3: "4564"
        },
        full_uac: "345365454564"
    },
    "100000999": {
        instrument_name: "dst2106a",
        case_id: "100000999",
        uac_chunks: {
            uac1: "9789",
            uac2: "7578",
            uac3: "5367"
        },
        full_uac: "978975785367"
    }
};

export const partialMatchedInstrumentUacDetails: InstrumentUacDetailsByCaseId = {
    "100000002": {
        instrument_name: "dst2106a",
        case_id: "100000002",
        uac_chunks: {
            uac1: "3453",
            uac2: "6545",
            uac3: "4564"
        },
        full_uac: "345365454564"
    }
};

export const emptyInstrumentUacDetails: InstrumentUacDetailsByCaseId = {};

export const validSampleFileWithUacArrayResponse: any[] = [
    {
        serial_number: "100000001",
        name: "Homer Simpson",
        phone: "5551234422",
        email: "homer@springfield.com",
        UAC1: "0009",
        UAC2: "7565",
        UAC3: "3827"
    },
    {
        serial_number: "100000002",
        name: "Seymour Skinner",
        phone: "1235663322",
        email: "a@b.c",
        UAC1: "3453",
        UAC2: "6545",
        UAC3: "4564"
    }
];

export const validSampleFileWithUacDatasResponse: Datas = [
    {
        serial_number: "100000001",
        name: "Homer Simpson",
        phone: "5551234422",
        email: "homer@springfield.com",
        UAC1: "0009",
        UAC2: "7565",
        UAC3: "3827"
    },
    {
        serial_number: "100000002",
        name: "Seymour Skinner",
        phone: "1235663322",
        email: "a@b.c",
        UAC1: "3453",
        UAC2: "6545",
        UAC3: "4564"
    }
];
