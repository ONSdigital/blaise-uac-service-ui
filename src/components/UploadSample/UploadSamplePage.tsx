import React, {ReactElement, useState} from "react";
import {Formik, Form} from "formik";
import {Link, Route, Switch} from "react-router-dom";
import App from "../../App";
import SelectFile from "./Sections/SelectFile";
import InstrumentName from "./Sections/InstrumentName";
import {ONSButton} from "blaise-design-system-react-components";
import {sampleFileAlreadyExists, generateUacCodesForSampleFile} from "../../client/file-functions";
import UploadFailed from "./Sections/UploadFailed";
import FileExists from "./Sections/FileExists";
import DownloadUacFile from "./Sections/DownloadUacFile";
import {Datas} from "react-csv-downloader/dist/esm/lib/csv";

function UploadSamplePage(): ReactElement {
    const [instrumentName, setInstrumentName] = useState<string>();
    const [overwrite, setOverwrite] = useState<string>();
    const [file, setFile] = useState<File>();
    const [activeStep, setActiveStep] = useState(0);
    const [downloadData, setDownloadData] = useState<Datas>([]);


    function _renderStepContent(step: number) {
        switch (step) {
            case 0:
                return (<InstrumentName instrumentName={instrumentName} setInstrumentName={setInstrumentName}/>);
            case 1:
                return (
                    <FileExists instrumentName={instrumentName} overwrite={overwrite} setOverwrite={setOverwrite}/>);
            case 2:
                return (<SelectFile file={file} setFile={setFile}/>);
            case 3:
                return (<DownloadUacFile instrumentName={instrumentName} downloadData={downloadData}/>);
            case 4:
                return (<UploadFailed instrumentName={instrumentName}/>);
        }
    }

    async function generateUacCodesAndSetDownloadData(): Promise<boolean> {
        const result = await generateUacCodesForSampleFile(instrumentName, file);
        setDownloadData(result);
        return result.length > 0;
    }

    async function _handleSubmit() {
        switch (activeStep) {
            case 0:
                setActiveStep(await sampleFileAlreadyExists(instrumentName) ? 1 : 2);
                break;
            case 1:
                setActiveStep(overwrite === "Yes" ? 2 : 3);
                break;
            case 2:
                setActiveStep(await generateUacCodesAndSetDownloadData() ? 3 : 4);
                break;
            default:
                setActiveStep(0);
        }
    }

    return (
        <>
            <main id="main-content" className="page__main u-mt-no">
                <Switch>
                    <Route path="/app">
                        <App/>
                    </Route>
                    <Route path="/">

                        <ul className="list list--bare list--inline u-mt-m">
                            <li className="list__item">
                                <Link to="/app" id="generate-uac-link">
                                    Home
                                </Link>
                            </li>
                        </ul>
                    </Route>
                </Switch>

                <Formik
                    validateOnBlur={false}
                    validateOnChange={false}
                    initialValues={{override: ""}}
                    onSubmit={_handleSubmit}>

                    <Form id={"formID"}>
                        <div>
                            <div className="u-mt-m">
                                {_renderStepContent(activeStep)}
                            </div>
                            <div className="btn-group u-mt-m">
                                <ONSButton
                                    id={"next-screen-button"}
                                    submit={true}
                                    primary={true} label={"Continue"}/>
                            </div>
                        </div>
                    </Form>

                </Formik>
            </main>
        </>
    );
}

export default UploadSamplePage;
