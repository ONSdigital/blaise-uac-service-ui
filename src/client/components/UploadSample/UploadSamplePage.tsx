import React, {ReactElement, useState} from "react";
import {Formik, Form} from "formik";
import {Link, Route, Switch} from "react-router-dom";
import App from "../../../App";
import SelectFile from "./Sections/SelectFile";
import InstrumentName from "./Sections/InstrumentName";
import {ONSPanel, ONSButton, StyledFormErrorSummary} from "blaise-design-system-react-components";
import {uploadFile} from "../../upload-file";
import UploadFailed from "./Sections/UploadFailed";
import UploadSuccessful from "./Sections/UploadSuccessful";

function UploadSamplePage(): ReactElement {
    const [instrumentName, setInstrumentName] = useState<string>();
    const [file, setFile] = useState<File>();
    const [activeStep, setActiveStep] = useState(0);

    function _renderStepContent(step: number) {
        switch (step) {
            case 1:
                return (
                    <UploadSuccessful instrumentName={instrumentName}/>
                );
            case 2:
                return (
                    <UploadFailed instrumentName={instrumentName}/>
                );
        }
    }

    async function _handleSubmit() {
        const result = await uploadFile(instrumentName, file);
        if(result === true) {
            setActiveStep(1);
        }
        else {
            setActiveStep(2);
        }
    }

    return (
        <>
            {/*<Breadcrumbs BreadcrumbList={*/}
            {/*    [*/}
            {/*        {link: "/", title: "Home"},*/}
            {/*    ]*/}
            {/*}/>*/}

            <main id="main-content" className="page__main u-mt-no">
                <Switch>
                    <Route path="/app">
                        <App/>
                    </Route>
                    <Route path="/">

                        <ul className="list list--bare list--inline u-mt-m">
                            <li className="list__item">
                                <Link to="/app" id="deploy-questionnaire-link">
                                    Home
                                </Link>
                            </li>
                        </ul>
                    </Route>
                </Switch>

                <Formik
                    validateOnBlur={false}
                    validateOnChange={false}
                    initialValues={{override: ""}}
                    onSubmit={_handleSubmit}>

                    <Form id={"formID"}>
                        <div>
                            <h1 className="u-mt-s">
                                Select instrument sample
                            </h1>

                            <ONSPanel>
                                <p>
                                    When a sample file is selected and you continue to upload this sample file, <b>this
                                    may take a few minutes</b>.
                                    <br/>
                                    <br/>
                                    Given this, <b>do not navigate away</b> from this page during this process. You will
                                    be
                                    re-directed
                                    when there is an update regarding the deploy of the questionnaire.
                                </p>
                            </ONSPanel>
                        </div>
                        <div className="u-mt-m">
                            <StyledFormErrorSummary/>
                            <InstrumentName
                                instrumentName={instrumentName}
                                setInstrumentName={setInstrumentName}/>
                            <SelectFile
                                file={file}
                                setFile={setFile}
                                loading={false}/>
                        </div>
                        {_renderStepContent(activeStep)}
                        <div className="btn-group u-mt-m">
                            <ONSButton
                                id={"upload-sample-button"}
                                submit={true}
                                primary={true} label={"Continue"}/>
                        </div>
                    </Form>

                </Formik>
            </main>
        </>
    );
}

export default UploadSamplePage;
