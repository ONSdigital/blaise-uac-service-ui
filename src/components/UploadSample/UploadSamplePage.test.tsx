import React from "react";
import {Router} from "react-router";
import {render, waitFor, fireEvent, cleanup, screen, act} from "@testing-library/react";
import "@testing-library/jest-dom";
import {createMemoryHistory} from "history";
import UploadSamplePage from "./UploadSamplePage";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import {getFileName} from "../../client/file-functions";
import {validSampleFileWithUacDatasResponse} from "../../mocks/csv-mocks";

const mock = new MockAdapter(axios, {onNoMatch: "throwException"});

const instrumentName = "DST1234A";
const fileName = getFileName(instrumentName);

describe("Upload Sample Page", () => {

    it("select file page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        await act(async () => await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        }));
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        const {queryByText} = render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        expect(queryByText(/Which questionnaire do you wish to generate Unique Access Codes for?/i)).toBeInTheDocument();
    });

    const invalidInstrumentNameTestCases = [
        {
            instrumentName: [
                null,
                undefined,
                "",
                "dst",
                "dst1",
                "2013dst",
                "ds2013a",
                "dst201a",
            ]
        }
    ];

    invalidInstrumentNameTestCases.forEach(test => {
        it(`Enter instrument name - should display an error message if you enter an invalid instrument name - ${test.instrumentName}`, async () => {
            const history = createMemoryHistory();
            render(
                <Router history={history}>
                    <UploadSamplePage/>
                </Router>
            );

            const inputInstrumentName = screen.getByLabelText(/questionnaire name/i);
            fireEvent.change(inputInstrumentName, {target: {value: test.instrumentName}});

            await fireEvent.click(screen.getByText(/Continue/));

            await waitFor(() => {
                expect(screen.queryAllByText("Enter a questionnaire name in the correct format, such as DST2004")).toHaveLength(2);
            });
        });
    });

    it("Enter instrument name - should navigate to the select file option if a sample for the instrument has not been previously uploaded", async () => {
        mock.onGet(`/api/v1/file/${fileName}/exists`).reply(200, false);

        await EnterInstrumentNameAndContinue();

        await waitFor(() => {
            expect(screen.queryAllByText("Select a sample file")).toHaveLength(1);
        });
    });

    it("Enter instrument name - should navigate to the overwrite sample option if a sample for the instrument has been previously uploaded", async () => {
        mock.onGet(`/api/v1/file/${fileName}/exists`).reply(200, true);

        await EnterInstrumentNameAndContinue();

        await waitFor(() => {
            expect(screen.queryAllByText("Overwrite sample file?")).toHaveLength(1);
        });
    });

    it("Overwrite sample - should navigate to the select file option if yes is selected", async () => {
        await NavigateToOverwriteFileAndContinue(true);

        await waitFor(() => {
            expect(screen.queryAllByText("Select a sample file")).toHaveLength(1);
        });
    });


    it("Select sample file - should display an error message if you dont select a file", async () => {
        await NavigateToSelectFile();

        await fireEvent.click(screen.getByText(/Continue/));

        await waitFor(() => {
            expect(screen.queryAllByText("Select a file")).toHaveLength(2);
        });
    });

    it("Select sample file - should display an error message if the selected file is not a .csv", async () => {
        await NavigateToSelectFileAndUpload("txt");

        await fireEvent.click(screen.getByText(/Continue/));

        await waitFor(() => {
            expect(screen.queryAllByText("File must be a .csv")).toHaveLength(2);
        });
    });

    it("Select sample file - should display an upload filed message if file fails to upload", async () => {
        mock.onPost(`/api/v1/instrument/${instrumentName}/uac/sample`).reply(500);

        await NavigateToSelectFileAndUpload("csv");

        await waitFor(() => {
            expect(screen.queryAllByText("File upload failed")).toHaveLength(1);
        });
    });

    it("Select sample file - should navigate to the download UAC option when a file is selected", async () => {
        mock.onPost(`/api/v1/instrument/${instrumentName}/uac/sample`).reply(201, validSampleFileWithUacDatasResponse);

        await NavigateToSelectFileAndUpload("csv");

        await waitFor(() => {
            expect(screen.queryAllByText(`Successfully generated UACs for ${instrumentName}`)).toHaveLength(1);
            expect(screen.queryAllByText("Download CSV file")).toHaveLength(1);
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

async function EnterInstrumentNameAndContinue() {
    const history = createMemoryHistory();
    render(
        <Router history={history}>
            <UploadSamplePage/>
        </Router>
    );

    //Enter instrument name
    const inputInstrumentName = screen.getByLabelText(/questionnaire name/i);
    fireEvent.change(inputInstrumentName, {target: {value: instrumentName}});

    await fireEvent.click(screen.getByText(/Continue/));
}

async function NavigateToOverwriteFileAndContinue(overwrite: boolean) {
    mock.onGet(`/api/v1/file/${fileName}/exists`).reply(200, true);

    await EnterInstrumentNameAndContinue();

    await waitFor(() => {
        expect(screen.queryAllByText("Overwrite sample file?")).toHaveLength(1);
    });

    const labelText = overwrite === true ? /Yes/i : /No/i;
    const overwriteSample = screen.getByLabelText(labelText);
    fireEvent.click(overwriteSample);

    await fireEvent.click(screen.getByText(/Continue/));
}

async function NavigateToSelectFile() {
    mock.onGet(`/api/v1/file/${fileName}/exists`).reply(200, false);

    await EnterInstrumentNameAndContinue();

    await waitFor(() => {
        expect(screen.queryAllByText("Select a sample file")).toHaveLength(1);
    });
}

async function NavigateToSelectFileAndUpload(fileExtension: string) {
    await NavigateToSelectFile();
    const inputEl = screen.getByLabelText(/Select a sample file/i);

    const file = new File(["(⌐□_□)"], `OPN2004A.${fileExtension}`, {
        type: fileExtension
    });

    Object.defineProperty(inputEl, "files", {
        value: [file]
    });

    fireEvent.change(inputEl);

    await fireEvent.click(screen.getByText(/Continue/));
}
