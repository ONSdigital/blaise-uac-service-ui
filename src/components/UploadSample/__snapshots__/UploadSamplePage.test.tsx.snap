// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Sample Page select file page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <ul
          class="list list--bare list--inline u-mt-m"
        >
          <li
            class="list__item"
          >
            <a
              href="/app"
              id="generate-uac-link"
            >
              Home
            </a>
          </li>
        </ul>
        <form
          action="#"
          id="formID"
        >
          <div>
            <div
              class="u-mt-m"
            >
              <div
                class="grid"
              >
                <div
                  class="grid__col col-8@m"
                >
                  <h1>
                    Which questionnaire do you wish to generate Unique Access Codes for?
                  </h1>
                  <div
                    class="panel panel--info panel--no-title  u-mt-m"
                  >
                    <div
                      class="panel__body "
                    >
                      <p>
                        Questionnaire name should in the formatted as three letters followed by four numbers, such as DST2004
                      </p>
                    </div>
                  </div>
                  <div
                    class="field"
                  >
                    <label
                      class="label "
                      for="questionnaire name"
                    >
                      Questionnaire name
                    </label>
                    <input
                      class="input input--text input-type__input"
                      id="questionnaire name"
                      name="questionnaire name"
                      type="input"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="btn-group u-mt-m"
            >
              <button
                class="btn "
                data-testid="button"
                id="next-screen-button"
                type="submit"
              >
                <span
                  class="btn__inner"
                >
                  Continue
                </span>
              </button>
            </div>
          </div>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="page__main u-mt-no"
      id="main-content"
    >
      <ul
        class="list list--bare list--inline u-mt-m"
      >
        <li
          class="list__item"
        >
          <a
            href="/app"
            id="generate-uac-link"
          >
            Home
          </a>
        </li>
      </ul>
      <form
        action="#"
        id="formID"
      >
        <div>
          <div
            class="u-mt-m"
          >
            <div
              class="grid"
            >
              <div
                class="grid__col col-8@m"
              >
                <h1>
                  Which questionnaire do you wish to generate Unique Access Codes for?
                </h1>
                <div
                  class="panel panel--info panel--no-title  u-mt-m"
                >
                  <div
                    class="panel__body "
                  >
                    <p>
                      Questionnaire name should in the formatted as three letters followed by four numbers, such as DST2004
                    </p>
                  </div>
                </div>
                <div
                  class="field"
                >
                  <label
                    class="label "
                    for="questionnaire name"
                  >
                    Questionnaire name
                  </label>
                  <input
                    class="input input--text input-type__input"
                    id="questionnaire name"
                    name="questionnaire name"
                    type="input"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn-group u-mt-m"
          >
            <button
              class="btn "
              data-testid="button"
              id="next-screen-button"
              type="submit"
            >
              <span
                class="btn__inner"
              >
                Continue
              </span>
            </button>
          </div>
        </div>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
